apache/metron
# METRON-1226 Searching Can Errantly Query the Wrong Indices

The ElasticsearchDAO can errantly query the wrong indices under certain conditions. This can occur when there are at least two different sensors that have been configured with a similar name prefix.

For example, consider the case when the user has created two different sensors; "snort" and "snort_v2".  When the user attempts to query the "snort" index in Elasticsearch, the query generated by the ElasticsearchDAO will hit both of these indices which is unexpected.

## Changes

Previously the `ElasticsearchDao` would append a simple `*` to the index name.  If the user had created a sensor "test" and another sensor "test_v2", then the wildcard logic would result in a pattern of `test*` when the user searches against the first sensor.  This pattern would unexpectedly also include data from the "test_v2" sensor.

The code was changed so that after wildcarding the index, it results in a pattern like `test_index*`.  Since "_index" is automatically appended by Metron, it can be used as a delimiter.  This will ensure that only the proper indices are searched.

**UPDATE:**  The name of the meta-alerts index was also changed from `metaalerts` to `metaalert_index`.  The new wildcard strategy would not work with the previous meta-alerts naming convention.  The simple way to address this was to rename the meta-alerts index.  This new name also more closely matches the way we name other indices.

## Testing

To manually test this specific condition, do the following in a Metron environment.
1. Configure a sensor called "test".
2. Allow telemetry to be consumed and indexed in Elasticsearch.
3. Create a second sensor called "test_v2".
4. Allow telemetry to be consumed and indexed in Elasticsearch.
5. Pass in a simple query such as the following.  Ensure that only data from the first sensor "test" is returned.  No data should be returned from "test_v2".
    ```
      {
       "indices": ["test"],
       "query": "*",
       "from": 4,
       "size": 3,
      "sort": [
        {
          "field": "timestamp",
         "sortOrder": "desc"
        }
      }
    ```
**UPDATE:**  The Alerts UI should also be tested to ensure there are no regressions.  I was able to perform queries and use the Alerts UI as before.

## Pull Request Checklist
